### ISSUE TITLE
Possible to cache results?

### ISSUE THREAD

**Issue by robertmclaws** [twitchax/AspNetCore.Proxy#55]
First off, thank you SO MUCH for this. I needed to proxy some calls to Webflow so that I could use their CMS to power some dynamic content in my app without having to build my own fractional CMS or redeploy every time I had changes to make.

Webflow has a rate-limit on their API, and I'm not expecting the results to change THAT often. Is there a built-in way to cache the results, or do I have to manually intercept the response and check it for changes? Would just adding a cache header to the response suffice?

Thanks!

**Comment by twitchax** [comment]
@robertmclaws, interesting.  As in, you want the proxy server itself to, effectively, memoize results that come back from the upstream?  And then, if a client makes the same call within some timeout, the memoized result is used?

**Comment by robertmclaws** [comment]
Yep! So let's say I have an API that only allows one request per second. Adding that option would let me set a 2 second cache, for example, to make sure I never exceed my rate limits.

You could also easily allow cache headers on the response, so the browser never makes the additional requests in the first place.

**Comment by szalapski** [comment]
Couldn't you just use the inside-controller proxy call and implement your
own caching around it?

**Comment by robertmclaws** [comment]
Yes, I could. But I think it should be a built-in feature.

**Comment by szalapski** [comment]
I'm having trouble seeing the benefit.

**Comment by twitchax** [comment]
@robertmclaws, that is an interesting idea: I will think about it.

@szalapski, do you think there is a purity reason it should not be done?  OTherwise, this sounds like something I would have fun doing when bored. :)

**Comment by szalapski** [comment]
Well, if the implementation of a feature can just as easily be implemented
AROUND a library call, why add code to enable the feature to be embedded in
it?

For example, do you want read-through, write-through, or write-behind
caching?  Do you want refresh-ahead?  How would expiration work?  Do you
really want to anticipate all these needs?

Better would be to ask, is there  any change needed to make it easier for
any developer to implement caching around the existing calls?  I can't
imagine what it would be, but I haven't tried it yet.

>

**Comment by twitchax** [comment]
Thanks, @szalapski, makes sense.

@robertmclaws, are there any features that you would need to help make this work?  I would think you could add some caching pretty easily in the `WithIntercept` options.

Basically, something like this.

```csharp
public class MyController : Controller
{
    private HttpProxyOptions _httpOptions = HttpProxyOptionsBuilder.Instance
        .WithIntercept(async context =>
        {
            if(_myCache.exists("blah) && /* other cache level requirements */)
            {
                var (status, response) = _myCache.get("blah");
                c.Response.StatusCode = status;
                await c.Response.WriteAsync(response /* or something that works here */);
                return true;
            }

            return false;
        }).Build();
    
    private MyCache _myCache = ...;
    
    [Route("api/posts/{postId}")]
    public Task GetPosts(int postId)
    {
        return this.HttpProxyAsync("http://myhttpendpoint.com", _httpOptions);
    }
}
```

**Comment by szalapski** [comment]
Would it be easier to implement caching in a controller action? Just
another option.

**Comment by twitchax** [comment]
That works, too.

**Comment by paulfryer** [comment]
+1 on providing cache features in the framework. I'm trying to do this now, it's complicated, mainly because I have to convert from an HttpResponseMessage that I get in the "WithAfterReceive" event to a HttpResponse when I fetch it from cache in the "WithIntercept" event.
